### PR TITLE
Event clarification and also guidance on not defaulting to the DRY principle

### PR DESCRIPTION
**Pull request by einari**
This should clarify some of the questions raised in #133.

Its also related to #205 - for more details and more to come.

### CODE REVIEW

**Comment by roarfred** [comment]
About events and indempotecy... does it mean the same event should be allowed to be raised several times, successivly, or does it (in addition) mean that you should allow the full event store to be replayed from any point to the end?

**Comment by einari** [comment]
You don't have to emit an event more than once. For idempotency it is important that when replaying you are able to reach the same result as you did when it was running. This affects how you model the event, but even more how you implement the processor

**Comment by karolikl** [comment]
I'd like to rephrase Roars question a bit: What should happen IF we receive the same event twice (let's say because of an error in the infrastructore? 

Example: We receive the TextMessageReceived event, which stores the text message, parses it and creates a case report. If we receive the same event again, should we: 
1. store the message again, process it again and create a new case report (leaving us with a duplicate)
or should we
2. identify that this event has already been processed and update the existing case report?

**Comment by einari** [comment]
At runtime you won't get the event twice. The replay is typically used when targeting a new read models repository. For event processors that would generate an email, sms or similar you'd implement a different method. Will announce that when is ready. 

But rule of thumb ; events won't happen  twice - it will be guaranteed once and once only. Ordering will also be guaranteed. 

In other words this is not something to worry about. 

Should I clarify this in the text somehow? It's kinda in there if you read it all. Reorder it? Rearrange? Have a checklist at the top of things you should think about - or a table? Very open for input here :)

**Comment by roarfred** [comment]
What puzzled me was the requirement for idempotency, as I see this could play out in three different scenarios:
1) The same event is received twice, don't let it affect your state (data store)
2) Playback from start at the initial state, make sure you always end up at the same state
3) Playback from an arbitrary point in time, at the current state. Make sure your state is not changed when all events are re-processed.

The last one is most about "what if I receive an event telling me something has changed, and this something doesn't exist in my state/data store?". I assume we should always be able to survive, even if the incoming event doesn't seem to make any sense. In the end, there is no way to bite back at the sender, is there?

**Comment by einari** [comment]
I'll rephrase it completely to avoid confusion. Maybe even remove it.

Idempontency is often something referred to in event sourcing - but the underlying system pretty much takes the problem away. The point of the paragraph is to make one aware of the concept. 

Let this be open till I fix it :)

**Comment by roarfred** [comment]
My comments were never meant to be an obstacle for the pull request, just found a spot to ask questions. Please excuse my ramblings and let the pull request go through, it clearly adds value to the master branch without these questions answered!

**Comment by einari** [comment]
:) no worries - good to get the message clear. Don't want confusion.

**Comment by einari** [comment]
Hope the last commit made it a bit clearer. 
Great to have these kinds of discussions and get the quality of the documentation up - when writing its easy to get trapped and make a lot of assumptions.

File: Documentation/Architecture/events.md [hunk 1]
```diff
@@ -120,16 +120,32 @@ precision on using [primitive types](#Primitive Types).
...
-In addition to events being modelled correctly and the processor doing it right, the system
-must guarantee ordering of events when replaying and also that your [event processor](./event_processor.md)
-can assume that events are being delivered once and once only. This is built into the underlying
-system being used. You as a developer do not have to worry about this.
+This does not mean that you should be keeping full entities on the event - still follow the rule of
+[complexity](#Complexity) as mentioned and remember to model the actual state change.
+Be aware of this and model accordingly.
+
+Important to keep in mind that the snapshot of a state is a series of events, these together form
+the snapshot. This should reduce what you keep on them but as a whole they help guarantee the
```
Comment (roarfred) [inline Documentation/Architecture/events.md]: Seems like somethings missing here

**Comment by roarfred** [comment]
Great clarification! Put in a comment about something that might have fallen out, just an incomplete sentence

File: Documentation/Architecture/events.md (hunk 1 above)
Comment (einari) [reply Documentation/Architecture/events.md]: LOL..   Me fix..   Probably a good idea to finish off sentences..

**Pull request merged**